src/color.rs

```diff
@@ -150,335 +97,606 @@ impl<'de> Deserialize<'de> for RGBA {
+/// <alpha-value>
+/// https://w3c.github.io/csswg-drafts/css-color-4/#alpha-syntax
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub struct AlphaValue {
+    /// The value as a number in the range of 0 to 1.
+    pub number: f32,
+}
+
+impl AlphaValue {
+    /// Construct a new AlphaValue from a number in the range of 0 to 1.
+    pub fn new(number: f32) -> Self {
+        Self { number }
+    }
+}
 
-/// Color specified by lightness, a- and b-axis components.
+// A set of coordinates in the Lab rectangular coordinate system.
 #[derive(Clone, Copy, PartialEq, Debug)]
-#[repr(C)]
-pub struct Lab {
-    /// The lightness component.
-    pub lightness: f32,
-    /// The a-axis component.
-    pub a: f32,
-    /// The b-axis component.
-    pub b: f32,
-    /// The alpha component.
-    pub alpha: f32,
+pub struct LabCoords {
+    pub lightness: Option<f32>,
+    pub a: Option<f32>,
+    pub b: Option<f32>,
+    pub alpha: Option<AlphaValue>,
+}
+
+impl LabCoords {
+    pub fn new(
+        lightness: Option<f32>,
+        a: Option<f32>,
+        b: Option<f32>,
+        alpha: Option<AlphaValue>,
+    ) -> Self {
+        Self {
+            lightness,
+            a,
+            b,
+            alpha,
+        }
+    }
 }
 
-/// Color specified by lightness, a- and b-axis components.
+// A set of coordinates in the LCH cylindrical coordinate system.
 #[derive(Clone, Copy, PartialEq, Debug)]
-#[repr(C)]
-pub struct Oklab {
-    /// The lightness component.
-    pub lightness: f32,
-    /// The a-axis component.
-    pub a: f32,
-    /// The b-axis component.
-    pub b: f32,
-    /// The alpha component.
-    pub alpha: f32,
+pub struct LchCoords {
+    pub lightness: Option<f32>,
+    pub chroma: Option<f32>,
+    pub hue: Option<Hue>,
+    pub alpha: Option<AlphaValue>,
 }
 
-macro_rules! impl_lab_like {
-    ($cls:ident, $fname:literal) => {
-        impl $cls {
-            /// Construct a new Lab color format with lightness, a, b and alpha components.
-            pub fn new(lightness: f32, a: f32, b: f32, alpha: f32) -> Self {
-                Self {
-                    lightness,
-                    a,
-                    b,
-                    alpha,
-                }
-            }
+impl LchCoords {
+    pub fn new(
+        lightness: Option<f32>,
+        chroma: Option<f32>,
+        hue: Option<Hue>,
+        alpha: Option<AlphaValue>,
+    ) -> Self {
+        Self {
+            lightness,
+            chroma,
+            hue,
+            alpha,
         }
+    }
+}
 
-        #[cfg(feature = "serde")]
-        impl Serialize for $cls {
-            fn serialize<S>(&self, serializer: S) -> Result<serde::ser::Ok, serde::ser::Error>
-            where
-                S: Serializer,
-            {
-                (self.lightness, self.a, self.b, self.alpha).serialize(serializer)
-            }
+/// A color in the sRGB color space.
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum SrgbColor {
```

Comment:
Will this enum also have the other predefined color spaces? `rec2020`, `linear-srgb`, `display-p3`, etc.?

Reply:
I haven't looked too deeply into the other predefined color spaces to know for sure, but the idea was for each independent color space to have its own enum, with a variant for each way to map into it. (That's why there's `CielabColor` and `OklabColor`.)

So if some of the predefined color spaces map directly into sRGB, then the `SrgbColor` enum would be expanded to include them. But completely independent color spaces like Display P3 and ITU-R BT.2020-2 would get their own enums, and the `Color` enum would get new variants to accommodate them.

But that, of course, is future work.